File: javascript/packages/linter/src/rules/a11y-no-accesskey-attribute.ts
import { type ParseResult, type ParserOptions, type HTMLAttributeNode, getAttributeName } from "@herb-tools/core"

import { BaseRuleVisitor } from "./rule-utils.js"
import { ParserRule } from "../types.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"

class NoAccesskeyAttributeVisitor extends BaseRuleVisitor {
  visitHTMLAttributeNode(node: HTMLAttributeNode): void {
    if (getAttributeName(node) === "accesskey") {
      this.addOffense(
        "Avoid using the `accesskey` attribute. Inconsistencies between keyboard shortcuts and keyboard commands used by screen readers and keyboard-only users create accessibility complications.",
        node.location,
      )
    }

    super.visitHTMLAttributeNode(node)
  }
}

export class A11yNoAccesskeyAttributeRule extends ParserRule {
  static ruleName = "a11y-no-accesskey-attribute"
  static introducedIn = this.version("0.9.3")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoAccesskeyAttributeVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_no_accesskey_attribute.rs
use crate::utils::tag_utils::get_attribute_name;

use herb::nodes::HTMLAttributeNode;
use herb::Visitor;

rule_visitor!(NoAccesskeyAttributeVisitor);
define_parser_rule!(
  A11yNoAccesskeyAttributeRule,
  "a11y-no-accesskey-attribute",
  Warning,
  NoAccesskeyAttributeVisitor,
  enabled: false,
  parser_options: { action_view_helpers: true }
);

impl Visitor for NoAccesskeyAttributeVisitor {
  fn visit_html_attribute_node(&mut self, node: &HTMLAttributeNode) {
    if get_attribute_name(node).as_deref() == Some("accesskey") {
      self.add_offense(
        "Avoid using the `accesskey` attribute. Inconsistencies between keyboard shortcuts and keyboard commands used by screen readers and keyboard-only users create accessibility complications.",
        node.location.clone(),
      );
    }

    self.walk_html_attribute_node(node);
  }
}
